### PR TITLE
Adds aarch64 fmla vector by element instruction

### PR DESCRIPTION
**Pull request by vflouris**
This commit does not add the half-float version
of the instruction which is introduced in armv8.2

Signed-off-by: Vasilis Flouris <vflouris@ics.forth.gr>

### CODE REVIEW

**Comment by derekbruening** [comment]
run arm tests

**Comment by derekbruening** [comment]
> This commit does not add the half-float version

Did you mean to include the word "not"?

**Comment by vflouris** [comment]
> > This commit does not add the half-float version
> 
> Did you mean to include the word "not"?

Yes I do, This variant of the instruction is added in the armv8.2 and I don't have a platform to test it.
I mention it on the commit message since it was supported on a previous version of dynamorio
and then removed.
Mind you, this variant has a different pattern (encoding) and so it requires an additional entry in 
codec.txt

**Comment by derekbruening** [comment]
> > > This commit does not add the half-float version
> > 
> > 
> > Did you mean to include the word "not"?
> 
> Yes I do, This variant of the instruction is added in the armv8.2 and I don't have a platform to test it.
> I mention it on the commit message since it was supported on a previous version of dynamorio
> and then removed.
> Mind you, this variant has a different pattern (encoding) and so it requires an additional entry in
> codec.txt

The confusion is that the entire commit description consists of "This commit does not add..." without saying what it *does* do.  In general please duplicate the title line contents within the body.  @AssadHashmi will be able to fix this up when merging.

**Comment by vflouris** [comment]
> > > > This commit does not add the half-float version
> > > 
> > > 
> > > Did you mean to include the word "not"?
> > 
> > 
> > Yes I do, This variant of the instruction is added in the armv8.2 and I don't have a platform to test it.
> > I mention it on the commit message since it was supported on a previous version of dynamorio
> > and then removed.
> > Mind you, this variant has a different pattern (encoding) and so it requires an additional entry in
> > codec.txt
> 
> The confusion is that the entire commit description consists of "This commit does not add..." without saying what it _does_ do. In general please duplicate the title line contents within the body. @AssadHashmi will be able to fix this up when merging.

Fair point. It was a last minute addition, thinking it would provide information to the person looking through git history to understand why the vector by element for half floats isn't there. I didn't notice how  it overshadowed the title line.

File: suite/tests/api/dis-a64.txt [hunk 1]
```diff
@@ -1765,6 +1765,12 @@ fd7fffff : ldr    d31, [sp,#32760]        : ldr    +0x7ff8(%sp)[8byte] -> %d31
 0e33cfa7 : fmla v7.2s, v29.2s, v19.2s               : fmla   %d7 %d29 %d19 $0x02 -> %d7
 4e33cfa7 : fmla v7.4s, v29.4s, v19.4s               : fmla   %q7 %q29 %q19 $0x02 -> %q7
 4e73cfa7 : fmla v7.2d, v29.2d, v19.2d               : fmla   %q7 %q29 %q19 $0x03 -> %q7
+4fd11240 : fmla v0.2d, v18.2d, v17.d[0]             : fmla   %q0 %q17 %q17 $0x00 $0x03 -> %q0
```
Comment (AssadHashmi) [inline suite/tests/api/dis-a64.txt]: The second operand should be %q18

File: suite/tests/api/dis-a64.txt [hunk 2]
```diff
@@ -1765,6 +1765,12 @@ fd7fffff : ldr    d31, [sp,#32760]        : ldr    +0x7ff8(%sp)[8byte] -> %d31
...
+4fc4180b : fmla v11.2d, v0.2d, v4.d[1]              : fmla   %q11 %q0 %q4  $0x01 $0x03 -> %q11
+4f981382 : fmla v2.4s, v28.4s, v24.s[0]             : fmla   %q2 %q28 %q24 $0x00 $0x02 -> %q2
+4fb81343 : fmla v3.4s, v26.4s, v24.s[1]             : fmla   %q3 %q26 %q24 $0x01 %0x02 -> %q3
```
Comment (AssadHashmi) [inline suite/tests/api/dis-a64.txt]: The sd_sz parameter should be $0x02

Comment (vflouris) [reply suite/tests/api/dis-a64.txt]: I amended the commit, it should fix those two, thanks a lot!

**Comment by AssadHashmi** [comment]
> This variant of the instruction is added in the armv8.2 and I don't have a platform to test it.

If you add the half-precision version I can test the patch on a v8.2 board I have access to.

**Comment by AssadHashmi** [comment]
run arm tests

File: suite/tests/api/dis-a64.txt [hunk 3]
```diff
@@ -1765,6 +1765,12 @@ fd7fffff : ldr    d31, [sp,#32760]        : ldr    +0x7ff8(%sp)[8byte] -> %d31
 0e33cfa7 : fmla v7.2s, v29.2s, v19.2s               : fmla   %d7 %d29 %d19 $0x02 -> %d7
 4e33cfa7 : fmla v7.4s, v29.4s, v19.4s               : fmla   %q7 %q29 %q19 $0x02 -> %q7
 4e73cfa7 : fmla v7.2d, v29.2d, v19.2d               : fmla   %q7 %q29 %q19 $0x03 -> %q7
+4fd11240 : fmla v0.2d, v18.2d, v17.d[0]             : fmla   %q0 %q18 %q17 $0x00 $0x03 -> %q0
+4fc4180b : fmla v11.2d, v0.2d, v4.d[1]              : fmla   %q11 %q0 %q4  $0x01 $0x03 -> %q11
```
Comment (AssadHashmi) [inline suite/tests/api/dis-a64.txt]: I think the precommit test regression is caused by an extra space between %q4  and $0x01. Unfortunately the comparison is whitespace sensitive.

Comment (vflouris) [reply suite/tests/api/dis-a64.txt]: I removed the white space

**Comment by vflouris** [comment]
> > This variant of the instruction is added in the armv8.2 and I don't have a platform to test it.
> 
> If you add the half-precision version I can test the patch on a v8.2 board I have access to.

In which case I will gladly add this as a separate pattern (it has a different fixed opcode so it has to be so). But since I will have to generate the hex integral value of the command by hand, it will take a bit. Would it be acceptable to do it in a separate commit?

**Comment by AssadHashmi** [comment]
> > > This variant of the instruction is added in the armv8.2 and I don't have a platform to test it.
> > 
> > 
> > If you add the half-precision version I can test the patch on a v8.2 board I have access to.

**Comment by AssadHashmi** [comment]
Sorry about accidentally closing the PR!



> > > This variant of the instruction is added in the armv8.2 and I don't have a platform to test it.
> > 
> > 
> > If you add the half-precision version I can test the patch on a v8.2 board I have access to.
> 
> In which case I will gladly add this as a separate pattern (it has a different fixed opcode so it has to be so). But since I will have to generate the hex integral value of the command by hand, it will take a bit. Would it be acceptable to do it in a separate commit?

A separate commit is fine.

**Comment by AssadHashmi** [comment]
run arm tests

**Pull request merged**